zrml/prediction-markets/src/lib.rs

```diff
@@ -474,26 +472,71 @@ mod pallet {
                 }
             }
-            Self::remove_last_dispute_from_market_ids_per_dispute_block(&disputes, &market_id)?;
+
             Self::set_market_as_disputed(&market, &market_id)?;
             let market_dispute = MarketDispute { at: curr_block_num, by: who, outcome };
             <Disputes<T>>::try_mutate(market_id, |disputes| {
                 disputes.try_push(market_dispute.clone()).map_err(|_| <Error<T>>::StorageOverflow)
             })?;
-            // each dispute resets dispute_duration
-            let dispute_duration_ends_at_block =
-                curr_block_num.saturating_add(market.deadlines.dispute_duration);
-            <MarketIdsPerDisputeBlock<T>>::try_mutate(dispute_duration_ends_at_block, |ids| {
-                ids.try_push(market_id).map_err(|_| <Error<T>>::StorageOverflow)
-            })?;
-
             Self::deposit_event(Event::MarketDisputed(
                 market_id,
                 MarketStatus::Disputed,
                 market_dispute,
             ));
-            // TODO(#782): add court benchmark
-            Ok((Some(T::WeightInfo::dispute_authorized(num_disputes, CacheSize::get()))).into())
+            Ok((Some(T::WeightInfo::dispute_authorized())).into())
+        }
+
+        /// Resolve the market,
+        /// if the dispute mechanism was unable to come to a conclusion in a specified time.
+        ///
+        /// # Weight
+        ///
+        /// Complexity: `O(n)`, where `n` is the number of outstanding disputes.
+        #[pallet::weight(
+            T::WeightInfo::resolve_expired_mdm_authorized_categorical()
+                .max(T::WeightInfo::resolve_expired_mdm_authorized_scalar())
+        )]
+        #[transactional]
+        pub fn resolve_failed_mdm(
+            origin: OriginFor<T>,
+            #[pallet::compact] market_id: MarketIdOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            ensure_signed(origin)?;
+            let market = <zrml_market_commons::Pallet<T>>::market(&market_id)?;
+            ensure!(market.status == MarketStatus::Disputed, Error::<T>::InvalidMarketStatus);
+
+            let disputes = Disputes::<T>::get(market_id);
+
+            // TODO(#782): use multiple benchmarks paths for different dispute mechanisms
+            let _has_failed = match market.dispute_mechanism {
+                MarketDisputeMechanism::Authorized => {
+                    T::Authorized::has_failed(&disputes, &market_id, &market)?
+                }
+                MarketDisputeMechanism::Court => {
+                    T::Court::has_failed(&disputes, &market_id, &market)?
+                }
+                MarketDisputeMechanism::SimpleDisputes => {
+                    T::SimpleDisputes::has_failed(&disputes, &market_id, &market)?
+                }
+            };
+
+            // TODO (#918): benchmarks only reach the end when a dispute mechanism has failed
+            let _has_failed = true;
+
```

Comment:
Oh I am sorry. This is indeed a mistake by me. Thanks for bringing this up. The problem here is, that I benchmark this extrinsic and none of the above MDMs actually return true to reach the end of the extrinsic. That's why I used this before and would like to use it again:

```rust
#[cfg(feature = "runtime-benchmarks")]
let _has_failed = true;
```

Reply:
9aa93c41501312977292171937fdd1ead112a2f5

Reply:
How would you try to reach the end of the extrinsic for benchmarking this extrinsic, when no MDM allows it to reach the end?

Reply:
https://github.com/zeitgeistpm/zeitgeist/pull/862#discussion_r1053673329

Reply:
> How would you try to reach the end of the extrinsic for benchmarking this extrinsic, when no MDM allows it to reach the end?

I wouldn't benchmark this at all. Let's summarize: You add a `[cfg(feature = ...)]` so that you can benchmark an unreachable bit of code, and then you only call the benchmark in the unreachable code. Why?

The real problem here is the unreachable code. Since you're proceeding in a step-by-step fashion to implement ZIP-0, the best solution is to remove the unused code and later add it in. In fact, you can just uncomment the entire `resolve_failed_mdm`, can't you? If it _always_ fails, we don't really need it now.

Reply:
Added in https://github.com/zeitgeistpm/zeitgeist/issues/918#issuecomment-1371775652

ceddfa944e675656035e025b0e7ff3aed0e642fe